tools/topology/topology2/cavs-mixin-mixout-eqiir-dts-ssp.conf

```diff
@@ -46,6 +46,26 @@ Object.Pipeline {
 				copier_type	"SSP"
 				stream_name	"$HEADSET_CODEC_NAME"
 				node_type	$I2S_LINK_OUTPUT_CLASS
+				IncludeByKey.SSP_HS_FMT_24 {
+				"true"	{
+						Object.Base.output_audio_format [
+							{
+								out_bit_depth		32
+								out_valid_bit_depth	24
+								out_sample_type		$SAMPLE_TYPE_MSB_INTEGER
+								out_fmt_cfg		"$[($out_channels | ($out_valid_bit_depth * 256))]"
```

Comment:
I'm I missing something, does the codec DAI specification also need a change ("sample_bits 24"), so that we change what is sent over I2S:.?
@kv2019i , codec dai no need to change the sample bits 24 bits. Because the valid bit 24 bits can be in 32 container. Thanks for pointing out, I think the SOF FW currently set LSB_INTEGER and it will use sample type to choose format conversion. It makes no difference to the codec with MSB_INTEGER. Remove it, any concerns?

Reply:
I guess we don't need to overwrite `out_fmt_cfg` if we don't set out_sample_type here right? @kv2019i @ranj063

Reply:
Right, but I think the problem is what we want to achieve (see Pierre's comment above). While ALH (for which we have tplg examples for 24bit output) takes MSB-format 32bit values, SSP behaves differently. We don't have any examples for 24bit SSP in upstream, so this is not so obvious, but SSP hw requires 24bit in the lower bits, so if the valid bits is 24bits in the SSP configuration, then copier needs to be configured to SAMPLE_TYPE_LSB_INTEGER here, and SSP configuration needs to set to have a  24bit configuration. 

DSP/Copier will send 32bit containers with audio in the lower 24bit, SSP will send the 24bit over SSP bus.

But yeah, @macchian , is above the correct configuration? The current cavs-rt5682.conf only defines 32bit SSP config. And if that is correct, then you don't need anything special in copier configuration. It outputs 32/32bit by default (MSB/LSB makes no differences if valid_bits==container_bits).

Reply:
@kv2019i , 
> But yeah, @macchian , is above the correct configuration?
yes, correct! It's what we'd like to do. As the exact as you mentioned, we want to achieve valid 24 bits / 32 bits container support.
"DSP/Copier will send 32bit containers with audio in the lower 24bit, SSP will send the 24bit over SSP bus."

Reply:
@macchian Right, so then your patch is correct but the sample type needs to be "out_sample_type		$SAMPLE_TYPE_LSB_INTEGER". This is very confusing, but hopefully having more examples will help. The reason for confusion stems from fact that historically the FW has "fixed up" some bad configurations. See e.g. commit 383d17a19e25f804d2ba37b6b621131918544e6a . So even if host/topology sets copier to input/output 24bit in a certain format, the FW has based on the link type (.e.g whether it's SSP/DMIC/ALH) overwritten the configuration and done the right thing. There are some legacy cases FW still needs to support, but we are trying to get rid of this in topology and have the correct formats in topology and not rely on the FW to fixup the values. This leads to some short term confusion, but in longer term, this will be less confusing for everybody -- tplg authoers and FW maintainers.

Reply:
> @macchian Right, so then your patch is correct but the sample type needs to be "out_sample_type $SAMPLE_TYPE_LSB_INTEGER".

@kv2019i , yes, it's clear to me. Add in/out_sample_type $SAMPLE_TYPE_LSB_INTEGER for easier review by everyone.